backend/src/worker.rs

```diff
@@ -1351,175 +1361,228 @@ async fn handle_child(
                 }
-            } => {
-                child.kill().await?;
-                return Err(Error::ExecutionErr("execution interrupted".to_string()).into())
             }
         };
-        r
-    });
-
-    let (tx, mut rx) = mpsc::channel::<String>(100);
-
-    tokio::spawn(async move {
-        while !done4.load(Ordering::Relaxed) {
-            let send = tokio::select! {
-                Ok(Some(out)) = reader.next_line() => {
-                    if out.len() > MAX_LOG_SIZE as usize {
-                        tracing::info!("Line is too big");
-                        let _ = tx.send(format!("Line is too big")).await;
-                        done4.store(true, Ordering::Relaxed);
-                        break;
-                    } else {
-                        tx.send(out).await
+
+        /* send SIGKILL and reap child process */
+        let (_, kill) = futures::future::join(set_reason, child.kill()).await;
+        kill.map(|()| None)
+    };
+
+    /* a future that reads output from the child and appends to the database */
+    let lines = async move {
+        /* log_remaining is zero when output limit was reached */
+        let mut log_remaining = (MAX_LOG_SIZE as usize).saturating_sub(logs.chars().count());
+        let mut result = io::Result::Ok(());
+        let mut output = output;
+        let mut write: OptionFuture<_> = None.into();
+
+        while let Some(line) = output.by_ref().next().await {
+            /* after receiving a line, continue until some delay has passed
+             * _and_ the previous database write is complete */
+            let until_write = write.clone();
+            let until = sleep(write_logs_delay).then(|()| until_write);
+            let mut read_lines = ready(line)
+                .into_stream()
+                .chain(output.by_ref())
+                .take_until(until)
+                .boxed();
+
+            /* Read up until an error is encountered,
+             * handle log lines first and then the error...
+             *
+             * (This looks a bit nicer using try_for_each but the side-effects/capturing
+             * in FnMut closure seems impractical, _maybe_ a futures::Sink would work but
+             * I know next to nothing about that.) */
+            let mut joined = String::new();
+
+            while let Some(line) = read_lines.next().await {
+                match line {
+                    Ok(_) if log_remaining == 0 => (),
+                    Ok(line) => {
+                        append_with_limit(&mut joined, &line, &mut log_remaining);
+
+                        *last_line = line;
+
+                        if log_remaining == 0 {
+                            tracing::info!(%job_id, "Too many logs lines");
+                            let _ = set_too_many_logs.send(true);
+                            joined.push_str(&format!(
+                                "Too many logs lines. Limit is {MAX_LOG_SIZE} chars. Killing job."
```

Comment:
not from this PR but maybe we should reword as `Too many logs lines or result too large to serialize` since the fact that result is just another log line a bit implementation specific

Reply:
Changed to 

> Job logs or result reached character limit of {MAX_LOG_SIZE}; killing job.